src/handlers/token.ts

```diff
@@ -59,189 +68,242 @@ export const getTokenExpiration = (tokenUser: TokenUser | TokenUserL2) =>
-          ).bimap(
-            () => new Error("Error generating JWT Token"),
-            _ => ({ tokenUser, tokenStr: _ })
+        ? pipe(
+            getUserJwt(
+              config.JWT_TOKEN_PRIVATE_KEY,
+              tokenUser,
+              tokenExpiration,
+              config.JWT_TOKEN_ISSUER,
+              config.JWT_TOKEN_KID,
+              config.JWT_TOKEN_AUDIENCE
+            ),
+            TE.mapLeft(() => new Error("Error generating JWT Token")),
+            TE.map((_) => ({ tokenUser, tokenStr: _ }))
           )
-        : taskEither
-            .of<Error, string>(crypto.randomBytes(32).toString("hex"))
-            .chain(_ =>
-              setWithExpirationTask(
-                redisClient,
-                `${SESSION_TOKEN_PREFIX}${_}`,
-                JSON.stringify(tokenUser),
-                tokenExpiration
-              ).bimap(
-                () => new Error("Error storing Opaque Token"),
-                () => ({
+        : pipe(
+            TE.of<Error, string>(crypto.randomBytes(32).toString("hex")),
+            TE.chain((_) =>
+              pipe(
+                setWithExpirationTask(
+                  redisClient,
+                  `${SESSION_TOKEN_PREFIX}${_}`,
+                  JSON.stringify(tokenUser),
+                  tokenExpiration
+                ),
+                TE.mapLeft(() => new Error("Error storing Opaque Token")),
+                TE.map(() => ({
                   tokenStr: _,
-                  tokenUser
-                })
+                  tokenUser,
+                }))
               )
             )
-    );
+          )
+    )
+  );
 
 export const introspectHandler = async (
   req: express.Request,
   res: express.Response // first check if token is blacklisted
-) =>
-  await existsKeyTask(
-    redisClient,
-    `${SESSION_INVALIDATE_TOKEN_PREFIX}${req.body.token}`
-  )
-    .mapLeft(() => res.status(500).json("Cannot introspect token"))
-    .chain(
-      fromPredicate(
-        _ => _ === false,
+): Promise<express.Response> => {
+  const result = pipe(
+    existsKeyTask(
+      redisClient,
+      `${SESSION_INVALIDATE_TOKEN_PREFIX}${req.body.token}`
+    ),
+    TE.mapLeft(() => res.status(500).json("Cannot introspect token")),
+    TE.chain(
+      TE.fromPredicate(
+        (_) => _ === false,
         () =>
           res.status(403).json({
-            active: false
+            active: false,
           })
       )
-    )
-    .chain(
-      fromPredicate(
+    ),
+    TE.chain(
+      TE.fromPredicate(
         () => !config.ENABLE_JWT,
         () => res.status(200).json({ active: true })
       )
-    )
-    .chain(() =>
-      getRawTokenUserFromRedis(req.body.token, res)
-        .chain<TokenUser | TokenUserL2>(_ =>
-          TokenUserL2.is(_)
-            ? mapDecoding(TokenUserL2, _).mapLeft(e =>
+    ),
+    TE.chain<any, boolean, TokenUser | TokenUserL2>(() =>
+      pipe(getRawTokenUserFromRedis(req.body.token, res), (_) =>
```

Comment:
What's the type of `_`? 
![Screen Shot 2022-04-28 at 12 54 42](https://user-images.githubusercontent.com/265564/165739827-3a54aca2-6cac-4873-bfe1-507c53c5276f.png)

I think we're missing a `chain` here.

Reply:
solved by https://github.com/pagopa/hub-spid-login-ms/commit/98e74009555375cf8aa477b3b7062329da9a41c9.